File: src/main/java/com/yourssu/pikiland/domain/service/LogTruncator.java
package com.yourssu.pikiland.domain.service;

import org.springframework.stereotype.Component;
import java.util.*;
import java.util.regex.Pattern;

@Component
public class LogTruncator {
    private static final Pattern ANSI_ESCAPE = Pattern.compile("\\u001B(?:[@-Z\\\\-_]|\\[[0-?]*[ -/]*[@-~])");
    private static final Pattern PROGRESS_BAR = Pattern.compile("\\[[=#>-]+\\s*\\]\\s*\\d+[%/]");
    
    private static final List<String> ERROR_KEYWORDS = Arrays.asList(
        "error", "exception", "failed", "fatal", "traceback",
        "segfault", "panic", "killed", "timeout", "abort", "syntax", "denied"
    );

    public String cleanAnsi(String text) {
        if (text == null) return "";
        return ANSI_ESCAPE.matcher(text).replaceAll("");
    }

    public String cleanProgressBars(String text) {
        if (text == null) return "";
        return PROGRESS_BAR.matcher(text).replaceAll("[PROGRESS]");
    }

    public String truncateLogForAi(String logText, int maxLines) {
        if (logText == null) return "";
        
        String cleaned = cleanProgressBars(cleanAnsi(logText));
        String[] lines = cleaned.split("\\r?\\n");
        int totalLines = lines.length;

        if (totalLines <= maxLines) {
            return cleaned;
        }

        int headCount = (int) (maxLines * 0.15);
        int tailCount = maxLines - headCount;

        List<String> headLines = Arrays.asList(Arrays.copyOfRange(lines, 0, headCount));
        List<String> tailLines = Arrays.asList(Arrays.copyOfRange(lines, totalLines - tailCount, totalLines));
        
        // Middle lines
        String[] middleLines = Arrays.copyOfRange(lines, headCount, totalLines - tailCount);
        
        // Find indices containing error keywords in middle section
        List<Integer> errorIndices = new ArrayList<>();
        for (int i = 0; i < middleLines.length; i++) {
            String lineLower = middleLines[i].toLowerCase();
            for (String keyword : ERROR_KEYWORDS) {
                if (lineLower.contains(keyword)) {
                    errorIndices.add(i);
                    break;
                }
            }
        }

        // Merge overlapping context intervals (context size = 5)
        int contextSize = 5;
        List<int[]> intervals = new ArrayList<>();
        for (int idx : errorIndices) {
            int start = Math.max(0, idx - contextSize);
            int end = Math.min(middleLines.length, idx + contextSize + 1);

            if (intervals.isEmpty()) {
                intervals.add(new int[]{start, end});
            } else {
                int[] last = intervals.get(intervals.size() - 1);
                if (start <= last[1]) {
                    last[1] = Math.max(last[1], end);
                } else {
                    intervals.add(new int[]{start, end});
                }
            }
        }

        List<String> extraLines = new ArrayList<>();
        for (int[] interval : intervals) {
            extraLines.add("\n--- [Error Context Detected (Middle)] ---");
            for (int i = interval[0]; i < interval[1]; i++) {
                extraLines.add(middleLines[i]);
            }
            extraLines.add("----------------------------------------\n");
        }

        int skippedCount = totalLines - maxLines;
        String systemAlert = String.format("\n... [System Alert: Truncated - %d lines omitted] ...\n... [Normal logs omitted to fit AI context limit] ...\n", skippedCount);

        StringBuilder sb = new StringBuilder();
        sb.append(String.join("\n", headLines));
        sb.append(systemAlert);
        
        if (!extraLines.isEmpty()) {
            sb.append(String.join("\n", extraLines));
            sb.append(systemAlert);
        }
        
        sb.append(String.join("\n", tailLines));
        return sb.toString();
    }
}

File: src/domain/log-truncator.ts
const ANSI_ESCAPE = /\u001B(?:[@-Z\\-_]|\[[0-?]*[ -/]*[@-~])/g;
const PROGRESS_BAR = /\[[=#>-]+\s*]\s*\d+[%/]/g;

const ERROR_KEYWORDS = [
  "error",
  "exception",
  "failed",
  "failure",
  "fatal",
  "traceback",
  "segfault",
  "panic",
  "killed",
  "timeout",
  "abort",
  "syntax",
  "denied",
  "caused by",
  "assert"
];

export class LogTruncator {
  public cleanAnsi(text: string | null | undefined): string {
    if (!text) return "";
    return text.replace(ANSI_ESCAPE, "");
  }

  public cleanProgressBars(text: string | null | undefined): string {
    if (!text) return "";
    return text.replace(PROGRESS_BAR, "[PROGRESS]");
  }

  public truncateLogForAi(rawLog: string | null | undefined, maxLines: number = 300): string {
    if (!rawLog || rawLog.trim().length === 0) {
      return "";
    }

    const cleaned = this.cleanProgressBars(this.cleanAnsi(rawLog));
    const lines = cleaned.split(/\r?\n/);
    const totalLines = lines.length;

    if (totalLines <= maxLines) {
      return cleaned;
    }

    const headCount = Math.floor(maxLines * 0.15);
    const tailCount = maxLines - headCount;

    const headLines = lines.slice(0, headCount);
    const tailLines = lines.slice(totalLines - tailCount);
    const middleLines = lines.slice(headCount, totalLines - tailCount);

    const errorIndices: number[] = [];
    for (let i = 0; i < middleLines.length; i++) {
      const lineLower = middleLines[i]!.toLowerCase();
      if (ERROR_KEYWORDS.some(keyword => lineLower.includes(keyword))) {
        errorIndices.push(i);
      }
    }

    const contextSize = 5;
    const intervals: Array<[number, number]> = [];
    for (const idx of errorIndices) {
      const start = Math.max(0, idx - contextSize);
      const end = Math.min(middleLines.length, idx + contextSize + 1);

      if (intervals.length === 0) {
        intervals.push([start, end]);
      } else {
        const last = intervals[intervals.length - 1]!;
        if (start <= last[1]) {
          last[1] = Math.max(last[1], end);
        } else {
          intervals.push([start, end]);
        }
      }
    }

    const extraLines: string[] = [];
    for (const [start, end] of intervals) {
      extraLines.push("\n--- [Error Context Detected (Middle)] ---");
      for (let i = start; i < end; i++) {
        extraLines.push(middleLines[i]!);
      }
      extraLines.push("----------------------------------------\n");
    }

    const skippedCount = totalLines - maxLines;
    const systemAlert = `\n... [System Alert: Truncated - ${skippedCount} lines omitted] ...\n... [Normal logs omitted to fit AI context limit] ...\n`;

    let result = headLines.join("\n") + systemAlert;

    if (extraLines.length > 0) {
      result += extraLines.join("\n") + systemAlert;
    }

    result += tailLines.join("\n");
    return result;
  }
}

